grpc/shared_conn.go

```diff
@@ -18,68 +24,190 @@ type OnTrackCB func(tr *webrtc.TrackRemote, r *webrtc.RTPReceiver)
+) (googlegrpc.ClientStream, error) {
+	return sc.grpcConn.NewStream(ctx, desc, method, opts...)
 }
 
 // AddOnTrackSub adds an OnTrack subscription for the resource.
 func (sc *SharedConn) AddOnTrackSub(name resource.Name, onTrackCB OnTrackCB) {
-	sc.mu.Lock()
-	defer sc.mu.Unlock()
-	sc.resourceOnTrackCBs[name] = onTrackCB
+	sc.resOnTrackMu.Lock()
+	defer sc.resOnTrackMu.Unlock()
+	sc.resOnTrackCBs[name] = onTrackCB
 }
 
 // RemoveOnTrackSub removes an OnTrack subscription for the resource.
 func (sc *SharedConn) RemoveOnTrackSub(name resource.Name) {
-	sc.mu.Lock()
-	defer sc.mu.Unlock()
-	delete(sc.resourceOnTrackCBs, name)
+	sc.resOnTrackMu.Lock()
+	defer sc.resOnTrackMu.Unlock()
+	delete(sc.resOnTrackCBs, name)
+}
+
+// GrpcConn returns a gRPC capable client connection.
+func (sc *SharedConn) GrpcConn() *ReconfigurableClientConn {
+	return &sc.grpcConn
 }
 
-// PeerConn returns the PeerConnection.
+// PeerConn returns a WebRTC PeerConnection capable of sending video/audio data.
 func (sc *SharedConn) PeerConn() *webrtc.PeerConnection {
 	return sc.peerConn
 }
 
+// ResetConn replaces the underlying GrpcConnection object. It also re-initiatlizes the
+// PeerConnection that must be renegotiated.
+func (sc *SharedConn) ResetConn(conn rpc.ClientConn, moduleLogger logging.Logger) {
+	sc.grpcConn.ReplaceConn(conn)
+
+	sc.logger = moduleLogger.Sublogger("conn")
+
+	peerConn, err := webrtc.NewPeerConnection(webrtc.Configuration{})
+	if err != nil {
+		sc.logger.Warnw("Unable to create optional peer connection for module. Ignoring.", "err", err)
+		return
+	}
+
+	sc.peerConn = peerConn
+	sc.PeerConnReady, err = rpc.ConfigureForRenegotiation(peerConn, sc.logger.AsZap())
+	if err != nil {
+		sc.logger.Warnw("Unable to create optional renegotiation channel for module. Ignoring.", "err", err)
+		utils.UncheckedError(sc.peerConn.Close())
+		sc.peerConn = nil
+		return
+	}
+
+	sc.peerConn.OnTrack(func(trackRemote *webrtc.TrackRemote, rtpReceiver *webrtc.RTPReceiver) {
+		name, err := resource.NewFromString(trackRemote.StreamID())
+		if err != nil {
+			sc.logger.Errorw("StreamID did not parse as a ResourceName", "sharedConn", fmt.Sprintf("%p", sc), "streamID", trackRemote.StreamID())
+			return
+		}
+		sc.resOnTrackMu.Lock()
+		onTrackCB, ok := sc.resOnTrackCBs[name]
+		sc.resOnTrackMu.Unlock()
+		if !ok {
+			sc.logger.Errorw("Callback not found for StreamID", "sharedConn", fmt.Sprintf("%p", sc), "streamID", trackRemote.StreamID())
+			return
+		}
+		onTrackCB(trackRemote, rtpReceiver)
+	})
+}
+
+// GenerateEncodedOffer creates a WebRTC offer that's JSON + base64 encoded. If an error is
+// returned, `SharedConn.PeerConn` will return nil until a following `Reset`.
+func (sc *SharedConn) GenerateEncodedOffer() (string, error) {
+	success := false
+	defer func() {
+		// Only keep the `peerConn` pointer active if we believe the connection is useable.
+		if !success {
+			sc.peerConn = nil
+		}
+	}()
+
+	pc := sc.PeerConn()
+	if pc == nil {
+		return "", nil
```

Comment:
Shouldn't this return an error given we don't expect it to result in establishing a working peer connection & the caller is able to recover from the error?

Reply:
Good catch. I've changed it to return an error.

I'm a bit torn on whether the clause should even exist. I don't think it can be hit with the way things stand right now.